CI: go test ./pkg/datapath/loader fails: Link does not exist

This issue is not observed in CI, however I notice it when running the privileged tests locally in my Ubuntu 22.04 environment against the master branch:

```
$ sudo -E make tests-privileged                                                                                                                                                                                                    [182/26533]
[sudo] password for joe:                                                                                                                                                                                                                                                                   
PRIVILEGED_TESTS=true PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin:/home/joe/git/cilium/bpf CGO_ENABLED=0 go test -mod=vendor -tags=osusergo -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyAddress=https://consul:8443 -X github.com/cilium/c
ilium/pkg/kvstore.etcdDummyAddress=http://etcd:4002 -X github.com/cilium/cilium/pkg/datapath.DatapathSHA256=e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855" \                                                                                                            
        ./... -test.v -timeout 600s -json -covermode=count -coverprofile=coverage.out | tparse                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                           
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓                                                                                                                                                                                                                        
┃   FAIL  package: github.com/cilium/cilium/pkg/datapath/loader   ┃                                                                                                                                                                                                                        
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                           
--- FAIL: Test (13.48s)                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                           
level=info msg="Using autogenerated IPv4 allocation range" subsys=node v4Prefix=10.1.0.0/16                                                                                                                                                                                                
level=info msg="Using autogenerated IPv6 allocation range" subsys=node v6Prefix="f00d::a01:0:0:0/96"                                                                                                                                                                                       
level=error msg="Link does not exist" device=cilium_net error="Link not found" subsys=datapath-loader                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                           
----------------------------------------------------------------------                                                                                                                                                                                                                     
FAIL: loader_test.go:165: LoaderTestSuite.TestCompileAndLoadHostEndpoint                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                           
loader_test.go:179:                                                                                                                                                                                                                                                                        
    s.testCompileAndLoad(c, &hostEp)                                                                                                                                                                                                                                                       
loader_test.go:154:                                                                                                                                                                                                                                                                        
    c.Assert(err, IsNil)                                                                                                                                                                                                                                                                   
... value netlink.LinkNotFoundError = netlink.LinkNotFoundError{error:(*errors.errorString)(0xc0005a20d0)} ("Link not found")                                                                                                                                                              
                                                                                                                                                                                                                                                                                           
level=info msg="Compiled new BPF template" BPFCompilationTime=1.201441695s file-path=/tmp/cilium_test1977592565/templates/66e66027ec503c0d3bd84f6df8c887e45638cf0ab81cf7d82218b48841fc471d/bpf_lxc.o subsys=datapath-loader                                                                
level=info msg="Compiled new BPF template" BPFCompilationTime=1.124382949s file-path=/tmp/cilium_test1977592565/templates/049a251dd9f895fb71a76f6349fcd7b95b39232474711e5fe7aabc4fe93e6df0/bpf_lxc.o subsys=datapath-loader                                                                
level=info msg="Compiled new BPF template" BPFCompilationTime=1.201681338s file-path=/tmp/cilium_test741519073/templates/3d74155c77be515a600048efd2bfd2db3811a6019f3fa693573f34d6a929ea22/bpf_lxc.o subsys=datapath-loader                                                                 
level=info msg="Compiled new BPF template" BPFCompilationTime=1.527059091s file-path=/tmp/cilium_test741519073/templates/7938b4f4365fae566285006eb62aca0b9a0c0006909a0fb7e06c9dd0596b2c27/bpf_lxc.o subsys=datapath-loader                                                                 
level=info msg="Compiled new BPF template" BPFCompilationTime=1.687688608s file-path=/tmp/cilium_test741519073/templates/c1cc079b9189e6659deb1bd51235fdb2f8833daee728422a90349bcbc6678a81/bpf_lxc.o subsys=datapath-loader                                                                 
level=info msg="Compiled new BPF template" BPFCompilationTime=1.740012043s file-path=/tmp/cilium_test741519073/templates/312c7c9cd762c4d8942c6278c249dc0fe183c4c22fcca1b87a2d595d2361cb16/bpf_lxc.o subsys=datapath-loader                                                                 
level=info msg="Compiled new BPF template" BPFCompilationTime=1.823694022s file-path=/tmp/cilium_test741519073/templates/3d74155c77be515a600048efd2bfd2db3811a6019f3fa693573f34d6a929ea22/bpf_lxc.o subsys=datapath-loader                                                                 
level=info msg="Compiled new BPF template" BPFCompilationTime=2.284098425s file-path=/tmp/cilium_test741519073/templates/312c7c9cd762c4d8942c6278c249dc0fe183c4c22fcca1b87a2d595d2361cb16/bpf_lxc.o subsys=datapath-loader                                                                 
level=info msg="Compiled new BPF template" BPFCompilationTime=2.477713267s file-path=/tmp/cilium_test741519073/templates/3d74155c77be515a600048efd2bfd2db3811a6019f3fa693573f34d6a929ea22/bpf_lxc.o subsys=datapath-loader
level=info msg="Compiled new BPF template" BPFCompilationTime=2.486516728s file-path=/tmp/cilium_test741519073/templates/34dfc680d96948579bcf46e860aa7317c10968dcc2246fe3bed80ea1b14b1596/bpf_lxc.o subsys=datapath-loader
level=info msg="Compiled new BPF template" BPFCompilationTime=2.494995266s file-path=/tmp/cilium_test741519073/templates/c1cc079b9189e6659deb1bd51235fdb2f8833daee728422a90349bcbc6678a81/bpf_lxc.o subsys=datapath-loader
level=info msg="Compiled new BPF template" BPFCompilationTime=2.693360568s file-path=/tmp/cilium_test741519073/templates/7d904e5b0e063ace19cd292a4a7b45a7a79848adaedcbfb0bcf6ba4441515500/bpf_lxc.o subsys=datapath-loader
level=info msg="Compiled new BPF template" BPFCompilationTime=2.781194192s file-path=/tmp/cilium_test741519073/templates/7938b4f4365fae566285006eb62aca0b9a0c0006909a0fb7e06c9dd0596b2c27/bpf_lxc.o subsys=datapath-loader
level=info msg="Compiled new BPF template" BPFCompilationTime=2.849800176s file-path=/tmp/cilium_test741519073/templates/b477b8a9d9f82a8fdb1243c350c35fee79757a0fa6c0aac5577225a3cef6431f/bpf_lxc.o subsys=datapath-loader
level=info msg="Compiled new BPF template" BPFCompilationTime=3.071249398s file-path=/tmp/cilium_test741519073/templates/b0130490420ac62f0608003819a40206ad48cc027c0184986c988c2477ec379a/bpf_lxc.o subsys=datapath-loader
level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv6.conf.dummy9.forwarding sysParamValue=1                                  
level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.dummy9.forwarding sysParamValue=1                                  
level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.dummy9.rp_filter sysParamValue=0                                   
level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.dummy9.accept_local sysParamValue=1                                
level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.dummy9.send_redirects sysParamValue=0                              
OOPS: 9 passed, 1 FAILED
...
```

--- a/pkg/datapath/linux/config/config_test.go
+++ b/pkg/datapath/linux/config/config_test.go
@@ -33,6 +33,7 @@ import (
 	"github.com/cilium/cilium/pkg/maps/nodemap/fake"
 	"github.com/cilium/cilium/pkg/option"
 	"github.com/cilium/cilium/pkg/testutils"
+	"github.com/cilium/cilium/pkg/testutils/netns"
 )
 
 var (
@@ -117,11 +118,56 @@ func writeConfig(t *testing.T, header string, write writeFn) {
 	}
 }
 
-func TestPrivilegedWriteNodeConfig(t *testing.T) {
-	setupConfigSuite(t)
-	writeConfig(t, "node", func(w io.Writer, dp datapath.ConfigWriter) error {
-		return dp.WriteNodeConfig(w, &dummyNodeCfg)
+func setupCiliumDummyDevices(t *testing.T, ns *netns.NetNS) {
+	t.Helper()
+	require.NoError(t, ns.Do(func() error {
+		ciliumNet := &netlink.Dummy{LinkAttrs: netlink.LinkAttrs{Name: "cilium_net"}}
+		if err := netlink.LinkAdd(ciliumNet); err != nil {
+			return err
+		}
+		if err := netlink.LinkSetUp(ciliumNet); err != nil {
+			return err
+		}
+		ciliumHost := &netlink.Dummy{LinkAttrs: netlink.LinkAttrs{Name: "cilium_host"}}
+		if err := netlink.LinkAdd(ciliumHost); err != nil {
+			return err
+		}
+		if err := netlink.LinkSetUp(ciliumHost); err != nil {
+			return err
+		}
+		return nil
+	}))
+
+	t.Cleanup(func() {
+		_ = ns.Do(func() error {
+			err1 := netlink.LinkDel(&netlink.Dummy{LinkAttrs: netlink.LinkAttrs{Name: "cilium_net"}})
+			err2 := netlink.LinkDel(&netlink.Dummy{LinkAttrs: netlink.LinkAttrs{Name: "cilium_host"}})
+			if err1 != nil && err2 != nil {
+				return fmt.Errorf("failed to delete cilium_net and cilium_host: %w", errors.Join(err1, err2))
+			}
+			if err1 != nil {
+				return fmt.Errorf("failed to delete cilium_net: %w", err1)
+			}
+			if err2 != nil {
+				return fmt.Errorf("failed to delete cilium_host: %w", err2)
+			}
+			return nil
+		})
 	})
+}
+
+func TestPrivilegedWriteNodeConfig(t *testing.T) {
+	testutils.PrivilegedTest(t)
+	ns := netns.NewNetNS(t)
+	setupCiliumDummyDevices(t, ns)
+	err := ns.Do(func() error {
+		setupConfigSuite(t)
+		writeConfig(t, "node", func(w io.Writer, dp datapath.ConfigWriter) error {
+			return dp.WriteNodeConfig(w, &dummyNodeCfg)
+		})
+		return nil
+	})
+	require.NoError(t, err)
 }
 
 func TestPrivilegedWriteNetdevConfig(t *testing.T) {
@@ -235,80 +281,85 @@ return false;`, main1.Index, main2.Index), m)
 
 func TestPrivilegedWriteNodeConfigExtraDefines(t *testing.T) {
 	testutils.PrivilegedTest(t)
-	setupConfigSuite(t)
+	ns := netns.NewNetNS(t)
+	setupCiliumDummyDevices(t, ns)
+	err := ns.Do(func() error {
+		setupConfigSuite(t)
 
-	var (
-		na datapath.NodeAddressing
-	)
-	h := hive.New(
-		cell.Provide(
-			fakeTypes.NewNodeAddressing,
-		),
-		maglev.Cell,
-		cell.Invoke(func(
-			nodeaddressing datapath.NodeAddressing,
-		) {
-			na = nodeaddressing
-		}),
-	)
+		var (
+			na datapath.NodeAddressing
+		)
+		h := hive.New(
+			cell.Provide(
+				fakeTypes.NewNodeAddressing,
+			),
+			maglev.Cell,
+			cell.Invoke(func(
+				nodeaddressing datapath.NodeAddressing,
+			) {
+				na = nodeaddressing
+			}),
+		)
 
-	tlog := hivetest.Logger(t)
-	require.NoError(t, h.Start(tlog, context.TODO()))
-	t.Cleanup(func() { h.Stop(tlog, context.TODO()) })
+		tlog := hivetest.Logger(t)
+		require.NoError(t, h.Start(tlog, context.TODO()))
+		t.Cleanup(func() { h.Stop(tlog, context.TODO()) })
 
-	var buffer bytes.Buffer
+		var buffer bytes.Buffer
 
-	// Assert that configurations are propagated when all generated extra defines are valid
-	cfg, err := NewHeaderfileWriter(WriterParams{
-		NodeAddressing:   na,
-		NodeExtraDefines: nil,
-		NodeExtraDefineFns: []dpdef.Fn{
-			func() (dpdef.Map, error) { return dpdef.Map{"FOO": "0x1", "BAR": "0x2"}, nil },
-			func() (dpdef.Map, error) { return dpdef.Map{"BAZ": "0x3"}, nil },
-		},
-		Sysctl:  sysctl.NewDirectSysctl(afero.NewOsFs(), "/proc"),
-		NodeMap: fake.NewFakeNodeMapV2(),
+		// Assert that configurations are propagated when all generated extra defines are valid
+		cfg, err := NewHeaderfileWriter(WriterParams{
+			NodeAddressing:   na,
+			NodeExtraDefines: nil,
+			NodeExtraDefineFns: []dpdef.Fn{
+				func() (dpdef.Map, error) { return dpdef.Map{"FOO": "0x1", "BAR": "0x2"}, nil },
+				func() (dpdef.Map, error) { return dpdef.Map{"BAZ": "0x3"}, nil },
+			},
+			Sysctl:  sysctl.NewDirectSysctl(afero.NewOsFs(), "/proc"),
+			NodeMap: fake.NewFakeNodeMapV2(),
+		})
+		require.NoError(t, err)
+
+		buffer.Reset()
+		require.NoError(t, cfg.WriteNodeConfig(&buffer, &dummyNodeCfg))
+
+		output := buffer.String()
+		require.Contains(t, output, "define FOO 0x1\n")
+		require.Contains(t, output, "define BAR 0x2\n")
+		require.Contains(t, output, "define BAZ 0x3\n")
+
+		// Assert that an error is returned when one extra define function returns an error
+		cfg, err = NewHeaderfileWriter(WriterParams{
+			NodeAddressing:   fakeTypes.NewNodeAddressing(),
+			NodeExtraDefines: nil,
+			NodeExtraDefineFns: []dpdef.Fn{
+				func() (dpdef.Map, error) { return nil, errors.New("failing on purpose") },
+			},
+			Sysctl:  sysctl.NewDirectSysctl(afero.NewOsFs(), "/proc"),
+			NodeMap: fake.NewFakeNodeMapV2(),
+		})
+		require.NoError(t, err)
+
+		buffer.Reset()
+		require.Error(t, cfg.WriteNodeConfig(&buffer, &dummyNodeCfg))
+
+		// Assert that an error is returned when one extra define would overwrite an already existing entry
+		cfg, err = NewHeaderfileWriter(WriterParams{
+			NodeAddressing:   fakeTypes.NewNodeAddressing(),
+			NodeExtraDefines: nil,
+			NodeExtraDefineFns: []dpdef.Fn{
+				func() (dpdef.Map, error) { return dpdef.Map{"FOO": "0x1", "BAR": "0x2"}, nil },
+				func() (dpdef.Map, error) { return dpdef.Map{"FOO": "0x3"}, nil },
+			},
+			Sysctl:  sysctl.NewDirectSysctl(afero.NewOsFs(), "/proc"),
+			NodeMap: fake.NewFakeNodeMapV2(),
+		})
+		require.NoError(t, err)
+		buffer.Reset()
+		require.Error(t, cfg.WriteNodeConfig(&buffer, &dummyNodeCfg))
+		return nil
 	})
 	require.NoError(t, err)
-
-	buffer.Reset()
-	require.NoError(t, cfg.WriteNodeConfig(&buffer, &dummyNodeCfg))
-
-	output := buffer.String()
-	require.Contains(t, output, "define FOO 0x1\n")
-	require.Contains(t, output, "define BAR 0x2\n")
-	require.Contains(t, output, "define BAZ 0x3\n")
-
-	// Assert that an error is returned when one extra define function returns an error
-	cfg, err = NewHeaderfileWriter(WriterParams{
-		NodeAddressing:   fakeTypes.NewNodeAddressing(),
-		NodeExtraDefines: nil,
-		NodeExtraDefineFns: []dpdef.Fn{
-			func() (dpdef.Map, error) { return nil, errors.New("failing on purpose") },
-		},
-		Sysctl:  sysctl.NewDirectSysctl(afero.NewOsFs(), "/proc"),
-		NodeMap: fake.NewFakeNodeMapV2(),
-	})
-	require.NoError(t, err)
-
-	buffer.Reset()
-	require.Error(t, cfg.WriteNodeConfig(&buffer, &dummyNodeCfg))
-
-	// Assert that an error is returned when one extra define would overwrite an already existing entry
-	cfg, err = NewHeaderfileWriter(WriterParams{
-		NodeAddressing:   fakeTypes.NewNodeAddressing(),
-		NodeExtraDefines: nil,
-		NodeExtraDefineFns: []dpdef.Fn{
-			func() (dpdef.Map, error) { return dpdef.Map{"FOO": "0x1", "BAR": "0x2"}, nil },
-			func() (dpdef.Map, error) { return dpdef.Map{"FOO": "0x3"}, nil },
-		},
-		Sysctl:  sysctl.NewDirectSysctl(afero.NewOsFs(), "/proc"),
-		NodeMap: fake.NewFakeNodeMapV2(),
-	})
-	require.NoError(t, err)
-
-	buffer.Reset()
-	require.Error(t, cfg.WriteNodeConfig(&buffer, &dummyNodeCfg))
 }
 
 func TestPreferredIPv6Address(t *testing.T) {
@@ -383,31 +434,37 @@ func TestPreferredIPv6Address(t *testing.T) {
 
 func TestPrivilegedNewHeaderfileWriter(t *testing.T) {
 	testutils.PrivilegedTest(t)
-	setupConfigSuite(t)
+	ns := netns.NewNetNS(t)
+	setupCiliumDummyDevices(t, ns)
+	err := ns.Do(func() error {
+		setupConfigSuite(t)
 
-	a := dpdef.Map{"A": "1"}
-	var buffer bytes.Buffer
+		a := dpdef.Map{"A": "1"}
+		var buffer bytes.Buffer
 
-	_, err := NewHeaderfileWriter(WriterParams{
-		NodeAddressing:     fakeTypes.NewNodeAddressing(),
-		NodeExtraDefines:   []dpdef.Map{a, a},
-		NodeExtraDefineFns: nil,
-		Sysctl:             sysctl.NewDirectSysctl(afero.NewOsFs(), "/proc"),
-		NodeMap:            fake.NewFakeNodeMapV2(),
-	})
+		_, err := NewHeaderfileWriter(WriterParams{
+			NodeAddressing:     fakeTypes.NewNodeAddressing(),
+			NodeExtraDefines:   []dpdef.Map{a, a},
+			NodeExtraDefineFns: nil,
+			Sysctl:             sysctl.NewDirectSysctl(afero.NewOsFs(), "/proc"),
+			NodeMap:            fake.NewFakeNodeMapV2(),
+		})
 
-	require.Error(t, err, "duplicate keys should be rejected")
+		require.Error(t, err, "duplicate keys should be rejected")
 
-	cfg, err := NewHeaderfileWriter(WriterParams{
-		NodeAddressing:     fakeTypes.NewNodeAddressing(),
-		NodeExtraDefines:   []dpdef.Map{a},
-		NodeExtraDefineFns: nil,
-		Sysctl:             sysctl.NewDirectSysctl(afero.NewOsFs(), "/proc"),
-		NodeMap:            fake.NewFakeNodeMapV2(),
+		cfg, err := NewHeaderfileWriter(WriterParams{
+			NodeAddressing:     fakeTypes.NewNodeAddressing(),
+			NodeExtraDefines:   []dpdef.Map{a},
+			NodeExtraDefineFns: nil,
+			Sysctl:             sysctl.NewDirectSysctl(afero.NewOsFs(), "/proc"),
+			NodeMap:            fake.NewFakeNodeMapV2(),
+		})
+		require.NoError(t, err)
+		require.NoError(t, cfg.WriteNodeConfig(&buffer, &dummyNodeCfg))
+		require.Contains(t, buffer.String(), "define A 1\n")
+		return nil
 	})
 	require.NoError(t, err)
-	require.NoError(t, cfg.WriteNodeConfig(&buffer, &dummyNodeCfg))
-	require.Contains(t, buffer.String(), "define A 1\n")
 }
 
 var provideNodemap = cell.Provide(func() nodemap.MapV2 {

--- a/pkg/testutils/netns/netns.go
+++ b/pkg/testutils/netns/netns.go
@@ -9,8 +9,10 @@ import (
 	"github.com/cilium/cilium/pkg/netns"
 )
 
+type NetNS = netns.NetNS
+
 // NewNetNS returns a new network namespace.
-func NewNetNS(tb testing.TB) *netns.NetNS {
+func NewNetNS(tb testing.TB) *NetNS {
 	tb.Helper()
 
 	ns, err := netns.New()

